charmtools/build/builder.py

```diff
@@ -547,9 +547,30 @@ def normalize_outputdir(self):
                 od = repo
         elif ":" in od:
             od = od.basename
-        log.info("Composing into {}".format(od))
         self.output_dir = od
 
+    def _check_path(self, path_to_check):
+        home_dir = os.path.expanduser('~')
+        if home_dir == '~':
+            raise BuildError('Could not determine home directory')
+        return os.path.abspath(path_to_check).startswith(home_dir)
+
+    def check_paths(self):
+        paths_to_check = [
+            self.output_dir,
+            self.wheelhouse_overrides,
+            os.environ.get('JUJU_REPOSITORY'),
+            os.environ.get('LAYER_PATH'),
+            os.environ.get('INTERACE_PATH'),
+        ]
+        for path_to_check in paths_to_check:
+            if path_to_check and not self._check_path(path_to_check):
+                raise BuildError('Due to snap confinement, all paths must be '
```

Comment:
This isn't entirely true, `$SNAP_DATA_DIR` is also valid, though less likely to be used.